File: host/SecondDisplay.Host/Protocol.cs
using System.Buffers.Binary;
using System.Net.Sockets;

namespace SecondDisplay.Host;

public static class PacketType
{
    public const byte Hello = 0x01;
    public const byte Ready = 0x02;
    public const byte Video = 0x10;
    public const byte Cursor = 0x11;
    public const byte Touch = 0x20;
    public const byte Scroll = 0x21;
    public const byte Key = 0x22;
}

public static class Codec
{
    public const byte Jpeg = 0;
    public const byte H264 = 1;
    public const byte H265 = 2;
}

public record HelloPacket(uint Width, uint Height, uint Density, uint RefreshRate);
public record ReadyPacket(uint Width, uint Height, uint RefreshRate, byte CodecId);
public record TouchPacket(byte Action, byte PointerId, float X, float Y);
public record KeyPacket(byte Action, ushort KeyCode, uint MetaState, ushort ScanCode);

public static class Protocol
{
    public const int HeaderSize = 5; // 1 type + 4 length

    public static void WritePacket(NetworkStream stream, byte type, ReadOnlySpan<byte> payload)
    {
        Span<byte> header = stackalloc byte[HeaderSize];
        header[0] = type;
        BinaryPrimitives.WriteUInt32LittleEndian(header[1..], (uint)payload.Length);
        stream.Write(header);
        stream.Write(payload);
    }

    public static (byte type, byte[] payload) ReadPacket(NetworkStream stream)
    {
        Span<byte> header = stackalloc byte[HeaderSize];
        ReadExact(stream, header);
        byte type = header[0];
        uint length = BinaryPrimitives.ReadUInt32LittleEndian(header[1..]);
        if (length > 16 * 1024 * 1024)
            throw new InvalidDataException($"Packet too large: {length}");
        var payload = new byte[length];
        ReadExact(stream, payload);
        return (type, payload);
    }

    public static void WriteReady(NetworkStream stream, ReadyPacket pkt)
    {
        Span<byte> buf = stackalloc byte[13];
        BinaryPrimitives.WriteUInt32LittleEndian(buf[0..], pkt.Width);
        BinaryPrimitives.WriteUInt32LittleEndian(buf[4..], pkt.Height);
        BinaryPrimitives.WriteUInt32LittleEndian(buf[8..], pkt.RefreshRate);
        buf[12] = pkt.CodecId;
        WritePacket(stream, PacketType.Ready, buf);
    }

    public static void WriteVideoFrame(NetworkStream stream, long ptsMicros, bool keyframe, ReadOnlySpan<byte> data)
    {
        Span<byte> header = stackalloc byte[HeaderSize];
        header[0] = PacketType.Video;
        BinaryPrimitives.WriteUInt32LittleEndian(header[1..], (uint)(9 + data.Length));

        Span<byte> meta = stackalloc byte[9];
        BinaryPrimitives.WriteInt64LittleEndian(meta, ptsMicros);
        meta[8] = keyframe ? (byte)1 : (byte)0;

        stream.Write(header);
        stream.Write(meta);
        stream.Write(data);
    }

    public static void WriteCursor(NetworkStream stream, bool visible, int x, int y, int w, int h, ReadOnlySpan<byte> bgra)
    {
        Span<byte> header = stackalloc byte[HeaderSize];
        header[0] = PacketType.Cursor;
        BinaryPrimitives.WriteUInt32LittleEndian(header[1..], (uint)(17 + bgra.Length));

        Span<byte> meta = stackalloc byte[17];
        meta[0] = visible ? (byte)1 : (byte)0;
        BinaryPrimitives.WriteInt32LittleEndian(meta[1..], x);
        BinaryPrimitives.WriteInt32LittleEndian(meta[5..], y);
        BinaryPrimitives.WriteInt32LittleEndian(meta[9..], w);
        BinaryPrimitives.WriteInt32LittleEndian(meta[13..], h);

        stream.Write(header);
        stream.Write(meta);
        stream.Write(bgra);
    }

    public static HelloPacket ParseHello(ReadOnlySpan<byte> payload)
    {
        return new HelloPacket(
            BinaryPrimitives.ReadUInt32LittleEndian(payload[0..]),
            BinaryPrimitives.ReadUInt32LittleEndian(payload[4..]),
            BinaryPrimitives.ReadUInt32LittleEndian(payload[8..]),
            BinaryPrimitives.ReadUInt32LittleEndian(payload[12..])
        );
    }

    public static TouchPacket ParseTouch(ReadOnlySpan<byte> payload)
    {
        return new TouchPacket(
            payload[0],
            payload[1],
            BinaryPrimitives.ReadSingleLittleEndian(payload[2..]),
            BinaryPrimitives.ReadSingleLittleEndian(payload[6..])
        );
    }

    public static KeyPacket ParseKey(ReadOnlySpan<byte> payload)
    {
        // ScanCode is optional for backward compatibility with older clients (7-byte payload).
        ushort scanCode = payload.Length >= 9
            ? BinaryPrimitives.ReadUInt16LittleEndian(payload[7..])
            : (ushort)0;
        return new KeyPacket(
            payload[0],
            BinaryPrimitives.ReadUInt16LittleEndian(payload[1..]),
            BinaryPrimitives.ReadUInt32LittleEndian(payload[3..]),
            scanCode
        );
    }

    private static void ReadExact(NetworkStream stream, Span<byte> buffer)
    {
        int offset = 0;
        while (offset < buffer.Length)
        {
            int read = stream.Read(buffer[offset..]);
            if (read == 0) throw new IOException("Connection closed");
            offset += read;
        }
    }
}

File: android/SecondDisplay/app/src/main/java/com/seconddisplay/client/Protocol.kt
package com.seconddisplay.client

import java.io.InputStream
import java.io.OutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder

object PacketType {
    const val HELLO: Byte = 0x01
    const val READY: Byte = 0x02
    const val VIDEO: Byte = 0x10
    const val CURSOR: Byte = 0x11
    const val TOUCH: Byte = 0x20
    const val SCROLL: Byte = 0x21
    const val KEY: Byte = 0x22
}

data class ReadyPacket(
    val width: Int,
    val height: Int,
    val refreshRate: Int,
    val codec: Byte
)

data class VideoFrame(
    val ptsMicros: Long,
    val keyframe: Boolean,
    val data: ByteArray
)

data class CursorPacket(
    val visible: Boolean,
    val x: Int,
    val y: Int,
    val width: Int,
    val height: Int,
    val bgra: ByteArray
)

object Protocol {
    private const val HEADER_SIZE = 5
    private const val MAX_PACKET_SIZE = 16 * 1024 * 1024

    fun writeHello(out: OutputStream, width: Int, height: Int, density: Int, refreshRate: Int) {
        val payload = ByteBuffer.allocate(16).order(ByteOrder.LITTLE_ENDIAN)
            .putInt(width)
            .putInt(height)
            .putInt(density)
            .putInt(refreshRate)
            .array()
        writePacket(out, PacketType.HELLO, payload)
    }

    fun writeTouch(out: OutputStream, action: Byte, pointerId: Byte, x: Float, y: Float) {
        val payload = ByteBuffer.allocate(10).order(ByteOrder.LITTLE_ENDIAN)
            .put(action)
            .put(pointerId)
            .putFloat(x)
            .putFloat(y)
            .array()
        writePacket(out, PacketType.TOUCH, payload)
    }

    fun writeKey(out: OutputStream, action: Byte, keyCode: Int, metaState: Int, scanCode: Int = 0) {
        val payload = ByteBuffer.allocate(9).order(ByteOrder.LITTLE_ENDIAN)
            .put(action)
            .putShort(keyCode.toShort())
            .putInt(metaState)
            .putShort(scanCode.toShort())
            .array()
        writePacket(out, PacketType.KEY, payload)
    }

    fun readPacket(input: InputStream): Pair<Byte, ByteArray> {
        val header = readExact(input, HEADER_SIZE)
        val buf = ByteBuffer.wrap(header).order(ByteOrder.LITTLE_ENDIAN)
        val type = buf.get()
        val length = buf.getInt()
        if (length < 0 || length > MAX_PACKET_SIZE)
            throw IllegalStateException("Packet too large: $length")
        val payload = readExact(input, length)
        return Pair(type, payload)
    }

    fun parseReady(payload: ByteArray): ReadyPacket {
        val buf = ByteBuffer.wrap(payload).order(ByteOrder.LITTLE_ENDIAN)
        return ReadyPacket(
            width = buf.getInt(),
            height = buf.getInt(),
            refreshRate = buf.getInt(),
            codec = buf.get()
        )
    }

    fun parseVideoFrame(payload: ByteArray): VideoFrame {
        val buf = ByteBuffer.wrap(payload).order(ByteOrder.LITTLE_ENDIAN)
        val pts = buf.getLong()
        val flags = buf.get()
        val data = ByteArray(payload.size - 9)
        buf.get(data)
        return VideoFrame(pts, flags.toInt() and 1 != 0, data)
    }

    fun parseCursor(payload: ByteArray): CursorPacket {
        val buf = ByteBuffer.wrap(payload).order(ByteOrder.LITTLE_ENDIAN)
        val visible = buf.get().toInt() != 0
        val x = buf.getInt()
        val y = buf.getInt()
        val width = buf.getInt()
        val height = buf.getInt()
        val data = ByteArray(payload.size - 17)
        buf.get(data)
        return CursorPacket(visible, x, y, width, height, data)
    }

    private fun writePacket(out: OutputStream, type: Byte, payload: ByteArray) {
        val header = ByteBuffer.allocate(HEADER_SIZE).order(ByteOrder.LITTLE_ENDIAN)
            .put(type)
            .putInt(payload.size)
            .array()
        out.write(header)
        out.write(payload)
        out.flush()
    }

    private fun readExact(input: InputStream, size: Int): ByteArray {
        val buf = ByteArray(size)
        var offset = 0
        while (offset < size) {
            val read = input.read(buf, offset, size - offset)
            if (read < 0) throw IllegalStateException("Connection closed")
            offset += read
        }
        return buf
    }
}
